File: main/password_policy.py
# นโยบายรหัสผ่าน — นิยามกฎไว้ที่เดียว ใช้ทั้งฝั่ง server (บังคับจริง) และฝั่งหน้าเว็บ (checklist สด)
# หน้าเว็บดึงกฎชุดนี้ผ่าน GET /api/password-policy แล้วประเมินด้วยตรรกะเดียวกัน
# (static/js/password_policy.js) จึงไม่มีทางที่ checklist บนจอกับที่ server บังคับจะไม่ตรงกัน

import re

MIN_LENGTH = 8
MAX_LENGTH = 128

# kind ที่รองรับ (ทั้ง python และ js ทำเหมือนกัน):
#   length  ยาวอยู่ในช่วง min..max
#   regex   ต้อง match
RULES = [
    {
        "id": "length",
        "label": f"ยาวอย่างน้อย {MIN_LENGTH} ตัวอักษร",
        "kind": "length",
        "min": MIN_LENGTH,
        "max": MAX_LENGTH,
    },
    {"id": "lower",  "label": "มีตัวพิมพ์เล็ก (a-z)",         "kind": "regex", "pattern": "[a-z]"},
    {"id": "upper",  "label": "มีตัวพิมพ์ใหญ่ (A-Z)",         "kind": "regex", "pattern": "[A-Z]"},
    {"id": "digit",  "label": "มีตัวเลข (0-9)",               "kind": "regex", "pattern": "[0-9]"},
    {"id": "symbol", "label": "มีอักขระพิเศษ (!@#$%^&* ฯลฯ)", "kind": "regex", "pattern": "[^A-Za-z0-9]"},
]


def _rule_passed(rule: dict, password: str) -> bool:
    kind = rule["kind"]
    if kind == "length":
        return rule["min"] <= len(password) <= rule["max"]
    if kind == "regex":
        return re.search(rule["pattern"], password) is not None
    raise ValueError(f"ไม่รู้จักกฎชนิด {kind}")


def evaluate(password: str) -> list[dict]:
    # คืนสถานะของทุกข้อ (ไว้ให้ฝั่งเรียกเอาไปแสดงหรือทำ log)
    return [
        {"id": r["id"], "label": r["label"], "passed": _rule_passed(r, password or "")}
        for r in RULES
    ]


def failed_rules(password: str) -> list[dict]:
    return [r for r in evaluate(password) if not r["passed"]]


def policy_for_client() -> dict:
    # ส่งให้หน้าเว็บไปประเมินเอง (label + วิธีตรวจ) — ไม่มีความลับอะไรในนี้
    return {"min_length": MIN_LENGTH, "max_length": MAX_LENGTH, "rules": RULES}


def error_detail(failed: list[dict]) -> str:
    return "รหัสผ่านไม่ผ่านนโยบาย: " + " · ".join(r["label"] for r in failed)

File: main/static/js/password_policy.js
// ตรวจรหัสผ่านด้วย "กฎชุดเดียวกับฝั่ง server" — ดึงกฎมาจาก /api/password-policy ตอนเปิดหน้า
// (นิยามกฎอยู่ที่ main/password_policy.py ที่เดียว ที่นี่แค่ประเมินตามชนิดของกฎ)
window.PasswordPolicy = {
  loaded: false,
  rules: [],
  minLength: 8,

  async load() {
    if (this.loaded) return this
    try {
      const res = await fetch(window.APP_BASE + '/api/password-policy')
      if (!res.ok) throw new Error('โหลดนโยบายรหัสผ่านไม่สำเร็จ')
      const data = await res.json()
      this.rules = data.rules || []
      this.minLength = data.min_length || 8
      this.loaded = true
    } catch (e) {
      // โหลดไม่ได้ = ไม่แสดง checklist แต่ฝั่ง server ยังบังคับอยู่เหมือนเดิม
      this.loaded = false
    }
    return this
  },

  testRule(rule, password) {
    switch (rule.kind) {
      case 'length':
        return password.length >= rule.min && password.length <= rule.max
      case 'regex':
        return new RegExp(rule.pattern).test(password)
      default:
        return true
    }
  },

  // คืนสถานะทุกข้อไว้ให้หน้าเว็บ render เป็น checklist
  check(password) {
    const pw = password || ''
    return this.rules.map(r => ({ id: r.id, label: r.label, passed: this.testRule(r, pw) }))
  },

  allPassed(password) {
    const list = this.check(password)
    return list.length > 0 && list.every(r => r.passed)
  },
}
